### DISCUSSION TITLE
SSL_shutdown() gives SIGPIPE exception. How not to get SIGPIPE?

### DISCUSSION THREAD

**Discussion by rakeshcpp** [openssl/openssl#24543] [Q&A] [answered]
In my SSL client I am getting SIGPIPE exception while call to SSL_shutdown(). I checked documentation, but I didn't get any hint of why I got SIGPIPE. 
What check should I made bufore calling SSL_shutdown()?

I just knew that when SSL_get_error() gives SSL_ERROR_SSL or SSL_ERROR_SYSCALL, I should not call SSL_shutdown().

I am not clear about how to use SSL_SENT_SHUTDOWN and SSL_RECEIVED_SHUTDOWN with SSL_shutdown()?
Means if I get SSL_RECEIVED_SHUTDOWN, should I call SSL_shotdown or it will give SIGPIPE?
and similarly clear for SSL_SENT_SHUTDOWN.

Please clearify. Thank you.

**Comment by kroeckx** [comment, marked as answer]
SIGPIPE is generated when you try to write to a pipe or fifo, but there is no reader. I'm not sure where the pipe or fifo is, but it seems the server has closed the connection already.

It's best to ignore the SIGPIPE with signal(SIGPIPE, SIG_IGN). This will result in getting an EPIPE error instead of the SIGPIPE. You'll still get the error, but now can handle the error.

You should probably not check SSL_RECEIVED_SHUTDOWN, but check that all data has been received. It might be useful to set SSL_SENT_SHUTDOWN to be able to reuse the session if you can't call SSL_shutdown. Please see the SSL_shutdown() manpage. If you need more help please give more information about what the client and server do.
